### ISSUE TITLE
IPAM old pod IPs not released! - IP pool is overflowing.

### ISSUE THREAD

**Issue by sedatsencan** [projectcalico/calico#5339]
## Expected Behavior
It needs to release old pod ips.

## Current Behavior
Old dead pod ip's are not deleted and the ip pool is overflowing.

## Errors
`kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "a8bd154853bc05d1f066d25d14312fde780ce18477d59adad3e06e2cb773a5c6": failed to request 1 IPv4 addresses. IPAM allocated only 0`

`error: rpc error: code = Unknown desc = failed to get sandbox ip: check network namespace closed: remove netns: unlinkat /var/run/netns/cni-a97226df-f866-cd53-5403-184784fcb06c: device or resource busy`

## Context
In my k8s cluster (v1.20.7), I detected that calico does not release old pod ip's. Old dead pod ip's are not deleted and the ip pool is overflowing.

> ~]# calicoctl ipam show
> +----------+----------------+-----------+-------- -----+-----------+
> | GROUPING |  CIDR     | IPS TOTAL | IPS IN USE  | IPS FREE |
> +----------+----------------+-----------+-------- -----+-----------+
> | IP Pool | 10.233.64.0/18 | 16384 | 15381 (94%) | 1003 (6%) |
> +----------+----------------+-----------+-------- -----+-----------+

It seems to be using 15381 ip, but that's impossible. When I look at the details, I see that the ips of the dead pods are not released. In this case, it leads to IP address leak and potentially IP pool depletion. I have to manually release the IP pool. But after a while it gets full again.

I need to use --ip parameter to be able to release. I do this with a for loop to unused ip blocks, but it's still unmanageable.

> ~]# calicoctl ipam release --help
> Set the Calico datastore access information in the environment variables or
> supply details in a config file.
> 
> Usage:
>   calicoctl ipam release --ip=<IP> [--config=<CONFIG>]
> 
> Options:
>   -h --help             Show this screen.
>      --ip=<IP>          IP address to release.
>   -c --config=<CONFIG>  Path to the file containing connection configuration in
>                         YAML or JSON format.
>                         [default: /etc/calico/calicoctl.cfg]

Any ideas, what could cause this? Or is there a solution?

## Your Environment

* Calico version: **Image: quay.io/calico/cni:v3.17.4**
* Orchestrator version (e.g. kubernetes, mesos, rkt): **v1.20.7**
* Operating System and version: **CentOS 7**
* Calicoctl Version : Client Version:    v3.17.4

**Comment by fasaxc** [comment]
We have fixed various sources of IPAM leaks over time, including adding an IPAM garbage collector (in v3.18, IIRC).  We also added `calicoctl ipam check` to help diagnose and clean up leaks when they occur.  Please upgrade to v3.20+ to get the latest IPAM garbage collection and let us know if that solves the problem.

**Comment by sedatsencan** [comment]
For some in-house reasons I was only able to upgrade the calico version to v3.19.2. But I see that the overflow has stopped in this version as well. Moreover, I can see the problem ip's with the check command. 
Thanks. @fasaxc

**Comment by rptaylor** [comment]
We had a similar problem with calico 3.19.2.  I found a good way to release all leaked IPs:
```
sudo /usr/local/bin/calicoctl.sh datastore migrate lock
sudo /usr/local/bin/calicoctl.sh ipam check --show-problem-ips -o report.json
sudo /usr/local/bin/calicoctl.sh ipam release --from-report=report.json
sudo /usr/local/bin/calicoctl.sh datastore migrate unlock
```

**Comment by rptaylor** [comment]
This type of problem can likely be caused by this issue on k8s v1.21:  https://github.com/kubernetes/kubernetes/pull/102344

**Comment by geotransformer** [comment]
Had this when upgrade Calico from 3.22 to 3.24, and K8S from 1.23 to 1.24, and also docker migration to containerd

**Comment by fasaxc** [comment]
@geotransformer How did you install Calico?  Are you using the kubernetes datastore?  When using the kubernetes datastore, Calico should garbage-collect any leaked IP addresses (the component that does that is `calico-kube-controllers`, is that component running and healthy in your cluster?)

**Comment by geotransformer** [comment]
> @geotransformer How did you install Calico? Are you using the kubernetes datastore? When using the kubernetes datastore, Calico should garbage-collect any leaked IP addresses (the component that does that is `calico-kube-controllers`, is that component running and healthy in your cluster?)

@fasaxc  we do use the "datastore_type": "kubernetes". 

Actually in our upgrade, there is also a host OS upgrade. So the 1-node k8s cluster will be reinitialized with old ETCD data. After k8s re-initialized, we found the pods are stuck at container creating.

Even calico-kube-controllers was in ctuck state.

user@xxx-node1:~$ kubectl describe pod -n kube-system calico-kube-controllers-84c476996d-f762r 
Pv4 addresses: Assigned 0 out of 1 requested IPv4 addresses; No IPs available in pools: [20.20.2.0/24]
Warning FailedCreatePodSandBox 63s (x241 over 54m) kubelet (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "9e4b1822702db6a204e852b77a9e4bbd13335be87b35b152c5c38bd2a0ec84c6": plugin type="calico" failed (add): failed to request IPv4 addresses: Assigned 0 out of 1 requested IPv4 addresses; No IPs available in pools: [20.20.2.0/24]
